lib/schema.js

```diff
@@ -1,214 +1,111 @@
-const toLongForm = (type, def, key) => {
-  if (typeof def !== 'object' && longFormOnlyTypes.has(type)) {
-    throw new Error(
-      `Short string only definition of "${key}: '${type}'" is not supported`
-    );
-  }
-  if (type === 'enum') {
-    return Object.assign(def, { type, enum: def.enum });
-  } else if (type === 'array' || type === 'set') {
-    const value = def[type];
-    Reflect.deleteProperty(def, type);
-    return Object.assign(def, { type, value });
-  } else if (type === 'map' || type === 'object') {
-    const col = def[type];
-    Reflect.deleteProperty(def, type);
-    const key = Object.keys(col)[0];
-    const value = col[key];
-    return Object.assign(def, { type, key, value });
-  }
-  return { type };
-};
+const RESERVED = ['validate', 'parse', 'serialize', 'format'];
 
 class Schema {
-  constructor(name, raw, namespaces) {
-    const short = shorthand(raw) || raw.type;
-    const defs = short ? { value: raw } : raw;
-    this.name = name;
-    this.kind = short ? 'scalar' : 'struct';
+  constructor(name, raw, namespaces = []) {
+    Schema.KIND = KIND;
+    this.kind = 'struct';
     this.scope = 'local';
     this.store = 'memory';
     this.allow = 'write';
-    this.namespaces = new Set(namespaces);
     this.parent = '';
-    this.fields = {};
     this.indexes = {};
     this.references = new Set();
     this.relations = new Set();
-    this.validate = defs.validate || null;
-    this.format = defs.format || null;
-    this.parse = defs.parse || null;
-    this.serialize = defs.serialize || null;
-    this.preprocess(defs);
+    this.fields = {};
+    this.name = name;
+    this.namespaces = new Set(namespaces);
+    const prep = new Preprocessor(this, Schema);
+    this.preprocess(raw, prep);
+    this.options = {
+      validate: raw.validate || null,
+      format: raw.format || null,
+      parse: raw.parse || null,
+      serialize: raw.serialize || null,
+    };
+  }
+
+  validate(value, path) {
+    if (!this.options.validate) return;
+    try {
+      return harmonize(this.options.validate(value, path));
+    } catch (err) {
+      return [`Field "${path || this.name}" validation failed ${String(err)}`];
+    }
   }
 
-  preprocess(defs) {
-    const keys = Object.keys(defs);
-    const first = keys[0];
-    if (isFirstUpper(first)) {
-      const metadata = defs[keys.shift()];
-      this.kind = toLowerCamel(first);
-      this.scope = metadata.scope || 'application';
-      this.store = metadata.store || 'persistent';
-      this.allow = metadata.allow || 'write';
-      if (metadata.schema && metadata.fields) {
-        this.parent = metadata.schema;
-        const parent = this.findReference(this.parent);
-        for (const key of metadata.fields) {
-          defs[key] = parent.fields[key];
-          keys.push(key);
-        }
-      }
+  preprocess(raw, prep) {
+    const source = prep.parse(raw);
+    if (isType(source)) {
+      this.fields = source;
```

Comment:
Does this mean that `fields` is not longer just an object with fields? This would probably break some things, why do we need this?

Reply:
In case if schema is a single type, we don't need to keep useless field `{ value: { type: 'string' } }` we can store any type shorthand directly in fields and check by constructor name with `isType`

Reply:
Logs look like this:
```js
fields: Type {
    references: Set(1) { 'string' },
    relations: Set(0) {},
    required: true,
    type: 'string'
  },
```

Reply:
Answered